<a href="https://colab.research.google.com/github/yeahzee0421/DAP/blob/main/lecutre9.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Import Libraries

In [690]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

## Get the Data

We'll use the built in breast cancer dataset from Scikit Learn. We can get with the load function:

In [691]:
from sklearn.datasets import load_breast_cancer

In [692]:
cancer = load_breast_cancer()

The data set is presented in a dictionary form:

In [693]:
cancer.keys()
#data:input 

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename', 'data_module'])

We can grab information and arrays out of this dictionary to set up our data frame and understanding of the features:

In [694]:
print(cancer['DESCR'])

.. _breast_cancer_dataset:

Breast cancer wisconsin (diagnostic) dataset
--------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 569

    :Number of Attributes: 30 numeric, predictive attributes and the class

    :Attribute Information:
        - radius (mean of distances from center to points on the perimeter)
        - texture (standard deviation of gray-scale values)
        - perimeter
        - area
        - smoothness (local variation in radius lengths)
        - compactness (perimeter^2 / area - 1.0)
        - concavity (severity of concave portions of the contour)
        - concave points (number of concave portions of the contour)
        - symmetry
        - fractal dimension ("coastline approximation" - 1)

        The mean, standard error, and "worst" or largest (mean of the three
        worst/largest values) of these features were computed for each image,
        resulting in 30 features.  For instance, field 0 is Mean Radi

In [695]:
cancer['feature_names']

array(['mean radius', 'mean texture', 'mean perimeter', 'mean area',
       'mean smoothness', 'mean compactness', 'mean concavity',
       'mean concave points', 'mean symmetry', 'mean fractal dimension',
       'radius error', 'texture error', 'perimeter error', 'area error',
       'smoothness error', 'compactness error', 'concavity error',
       'concave points error', 'symmetry error',
       'fractal dimension error', 'worst radius', 'worst texture',
       'worst perimeter', 'worst area', 'worst smoothness',
       'worst compactness', 'worst concavity', 'worst concave points',
       'worst symmetry', 'worst fractal dimension'], dtype='<U23')

## Set up DataFrame

In [696]:
cancer_df = pd.DataFrame(cancer['data'], columns=cancer['feature_names'])
cancer_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 30 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   mean radius              569 non-null    float64
 1   mean texture             569 non-null    float64
 2   mean perimeter           569 non-null    float64
 3   mean area                569 non-null    float64
 4   mean smoothness          569 non-null    float64
 5   mean compactness         569 non-null    float64
 6   mean concavity           569 non-null    float64
 7   mean concave points      569 non-null    float64
 8   mean symmetry            569 non-null    float64
 9   mean fractal dimension   569 non-null    float64
 10  radius error             569 non-null    float64
 11  texture error            569 non-null    float64
 12  perimeter error          569 non-null    float64
 13  area error               569 non-null    float64
 14  smoothness error         5

In [697]:
cancer_df['diagnosis'] = cancer.target
cancer_df.head()
#target은 0과 1로 이루어져있음 
#보통 환자를 positive(1)

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0                 0.07871  ...          17.33           184.60      2019.0   
1                 0.05667  ...          23.41           158.80      1956.0   
2                 0.05999  ...          25.53           152.50      1709.0   
3                 0.09744  ...          26.50            98.87       567.7   
4                 0.05883  ...          16.67           152.20      1575.0   

   worst smoothness  worst compactness  worst concavity  worst concave points  \
0            0.1622             0.6656           0.7119                0.2654   
1            0.1238             0.1866           0.2416                0.1860   
2            0.1444             0.4245           0.4504                0.2430   
3            0.2098             0.8663           0.6869                0.2575   
4            0.1374             0.2050           0.4000                0.1625   

   worst symmetry  worst fractal dimension  diagnosis  
0          0.4601                  0.11890          0  
1          0.2750                  0.08902          0  
2          0.3613                  0.08758          0  
3          0.6638                  0.17300          0  
4          0.2364                  0.07678          0  

[5 rows x 31 columns]

## Train Test Split

In [698]:
from sklearn.model_selection import train_test_split

In [699]:
X = cancer_df.drop(columns=['diagnosis'])
y = cancer_df['diagnosis']

In [700]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.30, random_state=0)

# Train the Support Vector Classifier
**[TODO]** Train the SVM Classifier with `gamma='auto` using `X_train` and `y_train`.

In [701]:
from sklearn.svm import SVC

model = SVC(gamma='auto')

model.fit(X_train, y_train)

SVC(gamma='auto')

## Predictions and Evaluations

**[TODO]** Now let's predict using the trained model.

In [702]:
predictions = model.predict(X_test)

In [703]:
from sklearn.metrics import classification_report, confusion_matrix

In [704]:
print(confusion_matrix(y_test, predictions))

[[  0  63]
 [  0 108]]


In [705]:
print(classification_report(y_test, predictions))
#precision recall이 모두 0이므로 모두 다 암환자라고 분류해버림 

              precision    recall  f1-score   support

           0       0.00      0.00      0.00        63
           1       0.63      1.00      0.77       108

    accuracy                           0.63       171
   macro avg       0.32      0.50      0.39       171
weighted avg       0.40      0.63      0.49       171



/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


Woah! Notice that we are classifying everything into a single class! This means our model needs to have it parameters adjusted (it may also help to normalize the data).

We can search for parameters using a GridSearch!

# Gridsearch

Finding the right parameters (like what `C` or `gamma values` to use) is a tricky task!

But luckily, we can be a little lazy and just try a bunch of combinations and see what works best! This idea of creating a `grid` of parameters and just trying out all the possible combinations is called a `Gridsearch`, this method is common enough that Scikit-learn has this functionality built in with [`GridSearchCV`](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html)! The CV stands for cross-validation.

The `GridSearchCV` takes a dictionary that describes the parameters that should be tried and a model to train. The grid of parameters is defined as a dictionary, where the keys are the parameters and the values are the settings to be tested. 

In [706]:
#svm 사용시 parameter에 민감한 경우엔 이 작업이 필요함 
param_grid = {'C': [0.1, 1, 10, 100, 1000],
              'gamma': [1, 0.1, 0.01, 0.001, 0.0001],
              'kernel': ['rbf']} 

In [707]:
#k-cross 5번씩 조합 다르게하면서 돌려봄 

from sklearn.model_selection import GridSearchCV

grid = GridSearchCV(SVC(), param_grid, refit=True, verbose=3)

# May take awhile!
grid.fit(X_train,y_train)

Fitting 5 folds for each of 25 candidates, totalling 125 fits
[CV 1/5] END ........C=0.1, gamma=1, kernel=rbf;, score=0.625 total time=   0.0s
[CV 2/5] END ........C=0.1, gamma=1, kernel=rbf;, score=0.625 total time=   0.0s
[CV 3/5] END ........C=0.1, gamma=1, kernel=rbf;, score=0.625 total time=   0.0s
[CV 4/5] END ........C=0.1, gamma=1, kernel=rbf;, score=0.633 total time=   0.0s
[CV 5/5] END ........C=0.1, gamma=1, kernel=rbf;, score=0.620 total time=   0.0s
[CV 1/5] END ......C=0.1, gamma=0.1, kernel=rbf;, score=0.625 total time=   0.0s
[CV 2/5] END ......C=0.1, gamma=0.1, kernel=rbf;, score=0.625 total time=   0.0s
[CV 3/5] END ......C=0.1, gamma=0.1, kernel=rbf;, score=0.625 total time=   0.0s
[CV 4/5] END ......C=0.1, gamma=0.1, kernel=rbf;, score=0.633 total time=   0.0s
[CV 5/5] END ......C=0.1, gamma=0.1, kernel=rbf;, score=0.620 total time=   0.0s
[CV 1/5] END .....C=0.1, gamma=0.01, kernel=rbf;, score=0.625 total time=   0.0s
[CV 2/5] END .....C=0.1, gamma=0.01, kernel=rbf

GridSearchCV(estimator=SVC(),
             param_grid={'C': [0.1, 1, 10, 100, 1000],
                         'gamma': [1, 0.1, 0.01, 0.001, 0.0001],
                         'kernel': ['rbf']},
             verbose=3)

You can inspect the best parameters found by GridSearchCV in the best_params_ attribute, and the best estimator in the best\_estimator_ attribute:

In [708]:
grid.best_params_

{'C': 1, 'gamma': 0.0001, 'kernel': 'rbf'}

In [709]:
grid.best_estimator_

SVC(C=1, gamma=0.0001)

Then you can re-run predictions on this grid object just like you would with a normal model.

In [710]:
grid_predictions = grid.predict(X_test)

In [711]:
print(confusion_matrix(y_test, grid_predictions))

[[ 59   4]
 [  3 105]]


In [712]:
print(classification_report(y_test,grid_predictions))

              precision    recall  f1-score   support

           0       0.95      0.94      0.94        63
           1       0.96      0.97      0.97       108

    accuracy                           0.96       171
   macro avg       0.96      0.95      0.96       171
weighted avg       0.96      0.96      0.96       171



# Todo 1

* 위의 실습에서 사용한 유방암 진단 데이터셋에 여러 모델을 적용해보고 모델의 성능을 accuracy, precision, recall, f1 score를 이용하여 확인하는 문제입니다.

### 1.1 다양한 분류기를 이용하여 분석 모델을 구축하기. 아래 모델들 중 자유롭게 선정하여 구현해보세요.
* [Logistic Regression](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html#)
* [kNN Classifier](https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsClassifier.html?highlight=kneigh#)
* [Naïve Bayes](https://scikit-learn.org/stable/modules/generated/sklearn.naive_bayes.GaussianNB.html?highlight=gaussiannb#)
* [Decision Tree](https://scikit-learn.org/stable/modules/tree.html)
* [Support Vector Machine](https://scikit-learn.org/stable/modules/generated/sklearn.svm.SVC.html?highlight=svc#)





decision tree 사용 시 hyperparameter 따로 지정해줘야됨 

In [713]:
#1. Logistic Regression

from sklearn.linear_model import LogisticRegression

# 모델 생성
model = LogisticRegression()

# 모델 훈련
model.fit(X_train, y_train)

# 예측 결과 구하기
y_predict = model.predict(X_test)

/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:818: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG,


### 1.2 생성한 모델의 성능 확인하기

* 분류에 대한 성능 지표(accuracy, precision, recall, f1 score)를 통해 생성한 모델의 성능을 확인한다.

In [714]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# model accuarcy
accuracy = accuracy_score(y_test, y_predict)

# model precision
precision = precision_score(y_test, y_predict)

# model recall
recall = recall_score(y_test, y_predict)

# model f1 score
f1 = f1_score(y_test, y_predict)

In [715]:
print("Accuracy: {0:.3f}".format(accuracy))
print("Precision: {0:.3f}".format(precision))
print("Recall: {0:.3f}".format(recall))
print("F1-Score: {0:.3f}".format(f1))

Accuracy: 0.953
Precision: 0.981
Recall: 0.944
F1-Score: 0.962


In [716]:
#2.kNN Classifier

from sklearn.neighbors import KNeighborsClassifier

model = KNeighborsClassifier(n_neighbors=7).fit(X_train, y_train)
y_predict = model.predict(X_test)

In [717]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# model accuarcy
accuracy = accuracy_score(y_test, y_predict)

# model precision
precision = precision_score(y_test, y_predict)

# model recall
recall = recall_score(y_test, y_predict)

# model f1 score
f1 = f1_score(y_test, y_predict)

print("Accuracy: {0:.3f}".format(accuracy))
print("Precision: {0:.3f}".format(precision))
print("Recall: {0:.3f}".format(recall))
print("F1-Score: {0:.3f}".format(f1))

Accuracy: 0.953
Precision: 0.963
Recall: 0.963
F1-Score: 0.963


# **나이브 베이즈 분류기**

Naive Bayes Classifier

텍스트 분류를 위해 전통적으로 사용되는 분류기.
feature끼리 서로 독립이라는 조건 필요.

In [718]:
cancer['feature_names']

array(['mean radius', 'mean texture', 'mean perimeter', 'mean area',
       'mean smoothness', 'mean compactness', 'mean concavity',
       'mean concave points', 'mean symmetry', 'mean fractal dimension',
       'radius error', 'texture error', 'perimeter error', 'area error',
       'smoothness error', 'compactness error', 'concavity error',
       'concave points error', 'symmetry error',
       'fractal dimension error', 'worst radius', 'worst texture',
       'worst perimeter', 'worst area', 'worst smoothness',
       'worst compactness', 'worst concavity', 'worst concave points',
       'worst symmetry', 'worst fractal dimension'], dtype='<U23')

In [719]:
#3.Naive Bayes

from sklearn.naive_bayes import GaussianNB

model = GaussianNB().fit(X_train, y_train)
y_predict = model.predict(X_test)

In [720]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# model accuarcy
accuracy = accuracy_score(y_test, y_predict)

# model precision
precision = precision_score(y_test, y_predict)

# model recall
recall = recall_score(y_test, y_predict)

# model f1 score
f1 = f1_score(y_test, y_predict)

print("Accuracy: {0:.3f}".format(accuracy))
print("Precision: {0:.3f}".format(precision))
print("Recall: {0:.3f}".format(recall))
print("F1-Score: {0:.3f}".format(f1))

Accuracy: 0.924
Precision: 0.944
Recall: 0.935
F1-Score: 0.940


# **결정 트리**

Yes or No로 이진분류 

각 feature의 전처리 작업 영향 크지 않음음

계속해서 분류 해나가면 무조건 학습 테스트 100% 나올 수 있지만 과적합 우려... ㅠㅠ 


PARAMETERS


*   max_depth = 트리 최대 깊이. 완벽하게 클래스 값이 결정될 때까지 분할. 또는 데이터 갯수가 min_sample_split보다 작아질 때까지 분할
깊이가 깊어지면 과적합 될 수 있으므로 제어 필요함.
*   min_samples_slipt = 노드를 분할하기 위한 최소한의 샘플 데이터 수. 과적합 제어하는데 사용. Default = 2, 작게 설정할 수록 분할 노드가 많아저 과적합 가능성 증가한다. 
*   max_feature = 최적의 분할을 위해 고려할 최대 feature 갯수 

*   random_state = 얘는 뭔지 도저히 이해가 안됐지만.. 쨌든 얘의의 값은 그렇게 중요하지 않고 None으로 두냐 정수를 넣느냐가 중요하다! 




In [721]:
#4. Decision Tree
from sklearn.tree import DecisionTreeClassifier

model = DecisionTreeClassifier(max_depth=4, random_state=1).fit(X_train, y_train)
y_predict = model.predict(X_test)

In [722]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# model accuarcy
accuracy = accuracy_score(y_test, y_predict)

# model precision
precision = precision_score(y_test, y_predict)

# model recall
recall = recall_score(y_test, y_predict)

# model f1 score
f1 = f1_score(y_test, y_predict)

print("Accuracy: {0:.3f}".format(accuracy))
print("Precision: {0:.3f}".format(precision))
print("Recall: {0:.3f}".format(recall))
print("F1-Score: {0:.3f}".format(f1))

Accuracy: 0.947
Precision: 0.971
Recall: 0.944
F1-Score: 0.958


**max_depth = 3, random_state = 1**

Accuracy: 0.942
Precision: 0.962
Recall: 0.944
F1-Score: 0.953


**max_depth = 3, random_state = 50**

Accuracy: 0.936
Precision: 0.953
Recall: 0.944
F1-Score: 0.949

**max_depth = 3, random_state = 150**

Accuracy: 0.936
Precision: 0.953
Recall: 0.944
F1-Score: 0.949

**max_depth = 4, random_state = 1**

Accuracy: 0.947
Precision: 0.971
Recall: 0.944
F1-Score: 0.958

In [723]:
from sklearn.model_selection import GridSearchCV

model = DecisionTreeClassifier().fit(X_train, y_train)
y_predict = model.predict(X_test)

param_grid ={
    'max_depth' : [1,2,3],
    'min_samples_split' : [2,3]
}

grid_tree = GridSearchCV(DecisionTreeClassifier(), param_grid= param_grid, cv=5, verbose = 3, refit = True)
grid_tree.fit(X_train, y_train)

scores_df = pd.DataFrame(grid_tree.cv_results_)
scores_df[['params', 'mean_test_score', 'rank_test_score', 
           'split0_test_score', 'split1_test_score', 'split2_test_score']]


Fitting 5 folds for each of 6 candidates, totalling 30 fits
[CV 1/5] END ..max_depth=1, min_samples_split=2;, score=0.912 total time=   0.0s
[CV 2/5] END ..max_depth=1, min_samples_split=2;, score=0.912 total time=   0.0s
[CV 3/5] END ..max_depth=1, min_samples_split=2;, score=0.887 total time=   0.0s
[CV 4/5] END ..max_depth=1, min_samples_split=2;, score=0.924 total time=   0.0s
[CV 5/5] END ..max_depth=1, min_samples_split=2;, score=0.873 total time=   0.0s
[CV 1/5] END ..max_depth=1, min_samples_split=3;, score=0.912 total time=   0.0s
[CV 2/5] END ..max_depth=1, min_samples_split=3;, score=0.912 total time=   0.0s
[CV 3/5] END ..max_depth=1, min_samples_split=3;, score=0.887 total time=   0.0s
[CV 4/5] END ..max_depth=1, min_samples_split=3;, score=0.924 total time=   0.0s
[CV 5/5] END ..max_depth=1, min_samples_split=3;, score=0.873 total time=   0.0s
[CV 1/5] END ..max_depth=2, min_samples_split=2;, score=0.912 total time=   0.0s
[CV 2/5] END ..max_depth=2, min_samples_split=2;,

params  mean_test_score  rank_test_score  \
0  {'max_depth': 1, 'min_samples_split': 2}         0.901994                5   
1  {'max_depth': 1, 'min_samples_split': 3}         0.901994                5   
2  {'max_depth': 2, 'min_samples_split': 2}         0.907089                4   
3  {'max_depth': 2, 'min_samples_split': 3}         0.909589                3   
4  {'max_depth': 3, 'min_samples_split': 2}         0.924652                1   
5  {'max_depth': 3, 'min_samples_split': 3}         0.922120                2   

   split0_test_score  split1_test_score  split2_test_score  
0             0.9125             0.9125             0.8875  
1             0.9125             0.9125             0.8875  
2             0.9125             0.9000             0.8875  
3             0.9125             0.9125             0.8875  
4             0.9250             0.9250             0.9125  
5             0.9250             0.9125             0.9250

In [724]:
print('best parameters : ', grid_tree.best_params_)
print('best score : ', grid_tree.best_score_)

best parameters :  {'max_depth': 3, 'min_samples_split': 2}
best score :  0.9246518987341773



# SVM
---


일반적으로 초평면(hyperPlane)을 사용하여 데이터 분리하게 되면 무한개의 초평면이 존재하는데, 훈련 관측치들로부터 가장 멀리 떨어진 분리 초평면인 최대 마진 초평면 선택해야함. 하지만 이때의 초평면은 해당 데이터에서만 높은 정확도를 가져오는 과적합의 위험이 있음. 
-> Maximal Margin Classifier


하지만 초평면이 존재하지 않는 데이터라면?

> Sort Margin ! -> SVC(Suppot Vector Classifier)
약간의 오류를 허용, 과적합 방지도 가능. 

SVM : 더 높은 차원의 공간에서 SVC를 적합. 그리고 이 SVC가 비선형 커널과 결합될 때 얻어지는 분류기


---

 **Cost, Gamma 파라미터 튜닝은 필수!!**


*   Cost : SVM 모델에 얼만큼 오차를 허용할 것인가? 
Cost가 클 수록 margin의 폭이 좁아진다. 

*   Gamma : SVM에서 초평면(hyperplane)이 아닌 커널과 관련 있는 수치. 가우시안 함수의 표준편차를 조정하는 인자, 큰 값을 가질수록 표준편차가 작다. 땨랴서 R에서 svm 함수 kernel 인자에 "radial"이 설정될 때 gamma 값을 조정한다. 일반 초평면의 경우 kernel 인자에 "linear"가 입력되고 이때는 그냥 cost 인자만 조정해주면 됨. 


---

**SVM 커널 종류**


*   linear : 선형 SVM
*   rbf : 가우시안 커널. 성능 우수해서 가장 많이 쓴다. 
*   sigmoid
*   poly 





In [725]:
#5. svm
from sklearn.svm import SVC

model=SVC(gamma='auto').fit(X_train, y_train)
y_predict = model.predict(X_test)

In [726]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# model accuarcy
accuracy = accuracy_score(y_test, y_predict)

# model precision
precision = precision_score(y_test, y_predict)

# model recall
recall = recall_score(y_test, y_predict)

# model f1 score
f1 = f1_score(y_test, y_predict)

print("Accuracy: {0:.3f}".format(accuracy))
print("Precision: {0:.3f}".format(precision))
print("Recall: {0:.3f}".format(recall))
print("F1-Score: {0:.3f}".format(f1))

Accuracy: 0.632
Precision: 0.632
Recall: 1.000
F1-Score: 0.774


In [727]:
from sklearn.model_selection import GridSearchCV

param_grid = {'C': [0.01, 0.1, 1, 10, 100, 1000],
              'gamma': [10, 1, 0.1, 0.01, 0.001, 0.0001],
              'kernel': ['rbf']} 
 
grid_SVM = GridSearchCV(SVC(), param_grid, refit=True, cv = 5, verbose=3)
grid_SVM.fit(X_train,y_train)
# scores_df = pd.DataFrame(grid_SVM.cv_results_)
# scores_df[['params', 'mean_test_score', 'rank_test_score', 
#            'split0_test_score', 'split1_test_score', 'split2_test_score']]

Fitting 5 folds for each of 36 candidates, totalling 180 fits
[CV 1/5] END ......C=0.01, gamma=10, kernel=rbf;, score=0.625 total time=   0.0s
[CV 2/5] END ......C=0.01, gamma=10, kernel=rbf;, score=0.625 total time=   0.0s
[CV 3/5] END ......C=0.01, gamma=10, kernel=rbf;, score=0.625 total time=   0.0s
[CV 4/5] END ......C=0.01, gamma=10, kernel=rbf;, score=0.633 total time=   0.0s
[CV 5/5] END ......C=0.01, gamma=10, kernel=rbf;, score=0.620 total time=   0.0s
[CV 1/5] END .......C=0.01, gamma=1, kernel=rbf;, score=0.625 total time=   0.0s
[CV 2/5] END .......C=0.01, gamma=1, kernel=rbf;, score=0.625 total time=   0.0s
[CV 3/5] END .......C=0.01, gamma=1, kernel=rbf;, score=0.625 total time=   0.0s
[CV 4/5] END .......C=0.01, gamma=1, kernel=rbf;, score=0.633 total time=   0.0s
[CV 5/5] END .......C=0.01, gamma=1, kernel=rbf;, score=0.620 total time=   0.0s
[CV 1/5] END .....C=0.01, gamma=0.1, kernel=rbf;, score=0.625 total time=   0.0s
[CV 2/5] END .....C=0.01, gamma=0.1, kernel=rbf

GridSearchCV(cv=5, estimator=SVC(),
             param_grid={'C': [0.01, 0.1, 1, 10, 100, 1000],
                         'gamma': [10, 1, 0.1, 0.01, 0.001, 0.0001],
                         'kernel': ['rbf']},
             verbose=3)

In [728]:
print('best parameters : ', grid_SVM.best_params_)
print('best score : ', grid_SVM.best_score_)

best parameters :  {'C': 1, 'gamma': 0.0001, 'kernel': 'rbf'}
best score :  0.9371518987341771


In [729]:
grid_pred = grid_SVM.predict(X_test)

In [730]:
print(confusion_matrix(y_test, grid_pred))

[[ 59   4]
 [  3 105]]


In [731]:
print(classification_report(y_test,grid_predictions))

              precision    recall  f1-score   support

           0       0.95      0.94      0.94        63
           1       0.96      0.97      0.97       108

    accuracy                           0.96       171
   macro avg       0.96      0.95      0.96       171
weighted avg       0.96      0.96      0.96       171



# Todo 2

와인의 화학 성분 데이터를 활용하여 여러 classification 중 원하는 모델을 자유롭게 사용하여 와인의 종류를 구분하고 생성한 모델의 accuracy를 확인하는 문제입니다.

## 2.1 데이터 로드 및 전처리

* Data를 원하는 방법으로 전처리 하세요.

In [732]:
from sklearn.datasets import load_wine

wine = load_wine()

## 2.2 데이터 탐색하고 Train_set, Test_set 나누기

* test_size = 0.3, random_state = 0

In [733]:
# 데이터 셋의 설명 확인하기
print(wine['DESCR'])

.. _wine_dataset:

Wine recognition dataset
------------------------

**Data Set Characteristics:**

    :Number of Instances: 178 (50 in each of three classes)
    :Number of Attributes: 13 numeric, predictive attributes and the class
    :Attribute Information:
 		- Alcohol
 		- Malic acid
 		- Ash
		- Alcalinity of ash  
 		- Magnesium
		- Total phenols
 		- Flavanoids
 		- Nonflavanoid phenols
 		- Proanthocyanins
		- Color intensity
 		- Hue
 		- OD280/OD315 of diluted wines
 		- Proline

    - class:
            - class_0
            - class_1
            - class_2
		
    :Summary Statistics:
    
    ============================= ==== ===== ======= =====
                                   Min   Max   Mean     SD
    ============================= ==== ===== ======= =====
    Alcohol:                      11.0  14.8    13.0   0.8
    Malic Acid:                   0.74  5.80    2.34  1.12
    Ash:                          1.36  3.23    2.36  0.27
    Alcalinity of Ash:            1

In [734]:
print(wine['feature_names'])
print(wine['target_names'])

['alcohol', 'malic_acid', 'ash', 'alcalinity_of_ash', 'magnesium', 'total_phenols', 'flavanoids', 'nonflavanoid_phenols', 'proanthocyanins', 'color_intensity', 'hue', 'od280/od315_of_diluted_wines', 'proline']
['class_0' 'class_1' 'class_2']


In [735]:
wine_df = pd.DataFrame(wine['data'], columns = wine['feature_names'])
wine_df.head()

alcohol  malic_acid   ash  alcalinity_of_ash  magnesium  total_phenols  \
0    14.23        1.71  2.43               15.6      127.0           2.80   
1    13.20        1.78  2.14               11.2      100.0           2.65   
2    13.16        2.36  2.67               18.6      101.0           2.80   
3    14.37        1.95  2.50               16.8      113.0           3.85   
4    13.24        2.59  2.87               21.0      118.0           2.80   

   flavanoids  nonflavanoid_phenols  proanthocyanins  color_intensity   hue  \
0        3.06                  0.28             2.29             5.64  1.04   
1        2.76                  0.26             1.28             4.38  1.05   
2        3.24                  0.30             2.81             5.68  1.03   
3        3.49                  0.24             2.18             7.80  0.86   
4        2.69                  0.39             1.82             4.32  1.04   

   od280/od315_of_diluted_wines  proline  
0                          3.92   1065.0  
1                          3.40   1050.0  
2                          3.17   1185.0  
3                          3.45   1480.0  
4                          2.93    735.0

In [736]:
wine_df.isnull().sum()

alcohol                         0
malic_acid                      0
ash                             0
alcalinity_of_ash               0
magnesium                       0
total_phenols                   0
flavanoids                      0
nonflavanoid_phenols            0
proanthocyanins                 0
color_intensity                 0
hue                             0
od280/od315_of_diluted_wines    0
proline                         0
dtype: int64

In [737]:
# from sklearn.impute import SimpleImputer
# imp_mean = SimpleImputer(missing_values=np.nan, strategy='most_frequent')
# imp_mean.fit(wine_df)
# X = imp_mean.transform(wine_df)
# pd.DataFrame(X, columns = wine_df.columns)

In [738]:
# from sklearn.impute import KNNImputer
# imputer = KNNImputer(n_neighbors=2, weights='uniform')
# imputer.fit_transform(wine_df.select_dtypes(include='number'))
# pd.DataFrame(imputer.fit_transform(wine_df.select_dtypes(include='number')),
#              columns = wine_df.select_dtypes(include='number').columns)

*참고한 velog..

[결측치 처리](https://velog.io/@jee-9/%EA%B2%B0%EC%B8%A1%EC%B9%98Nulls-%EC%B2%98%EB%A6%AC%EC%97%90-%EB%8C%80%ED%95%B4%EC%84%9C-Imputation-Single-MICE)

In [739]:
#MICE 

from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import IterativeImputer

imputer = IterativeImputer()
imputer.fit(wine_df)
pd.DataFrame(imputer.fit_transform(wine_df), columns = wine_df.columns)

alcohol  malic_acid   ash  alcalinity_of_ash  magnesium  total_phenols  \
0      14.23        1.71  2.43               15.6      127.0           2.80   
1      13.20        1.78  2.14               11.2      100.0           2.65   
2      13.16        2.36  2.67               18.6      101.0           2.80   
3      14.37        1.95  2.50               16.8      113.0           3.85   
4      13.24        2.59  2.87               21.0      118.0           2.80   
..       ...         ...   ...                ...        ...            ...   
173    13.71        5.65  2.45               20.5       95.0           1.68   
174    13.40        3.91  2.48               23.0      102.0           1.80   
175    13.27        4.28  2.26               20.0      120.0           1.59   
176    13.17        2.59  2.37               20.0      120.0           1.65   
177    14.13        4.10  2.74               24.5       96.0           2.05   

     flavanoids  nonflavanoid_phenols  proanthocyanins  color_intensity   hue  \
0          3.06                  0.28             2.29             5.64  1.04   
1          2.76                  0.26             1.28             4.38  1.05   
2          3.24                  0.30             2.81             5.68  1.03   
3          3.49                  0.24             2.18             7.80  0.86   
4          2.69                  0.39             1.82             4.32  1.04   
..          ...                   ...              ...              ...   ...   
173        0.61                  0.52             1.06             7.70  0.64   
174        0.75                  0.43             1.41             7.30  0.70   
175        0.69                  0.43             1.35            10.20  0.59   
176        0.68                  0.53             1.46             9.30  0.60   
177        0.76                  0.56             1.35             9.20  0.61   

     od280/od315_of_diluted_wines  proline  
0                            3.92   1065.0  
1                            3.40   1050.0  
2                            3.17   1185.0  
3                            3.45   1480.0  
4                            2.93    735.0  
..                            ...      ...  
173                          1.74    740.0  
174                          1.56    750.0  
175                          1.56    835.0  
176                          1.62    840.0  
177                          1.60    560.0  

[178 rows x 13 columns]

In [740]:
wine_df = pd.get_dummies(wine_df)
wine_df.head()

alcohol  malic_acid   ash  alcalinity_of_ash  magnesium  total_phenols  \
0    14.23        1.71  2.43               15.6      127.0           2.80   
1    13.20        1.78  2.14               11.2      100.0           2.65   
2    13.16        2.36  2.67               18.6      101.0           2.80   
3    14.37        1.95  2.50               16.8      113.0           3.85   
4    13.24        2.59  2.87               21.0      118.0           2.80   

   flavanoids  nonflavanoid_phenols  proanthocyanins  color_intensity   hue  \
0        3.06                  0.28             2.29             5.64  1.04   
1        2.76                  0.26             1.28             4.38  1.05   
2        3.24                  0.30             2.81             5.68  1.03   
3        3.49                  0.24             2.18             7.80  0.86   
4        2.69                  0.39             1.82             4.32  1.04   

   od280/od315_of_diluted_wines  proline  
0                          3.92   1065.0  
1                          3.40   1050.0  
2                          3.17   1185.0  
3                          3.45   1480.0  
4                          2.93    735.0

In [741]:
wine_df['wine'] = wine.target

In [742]:
# X, Y 설정
X = wine_df.drop(columns=['wine' ,'total_phenols'])
Y = wine_df['wine']

In [743]:
from sklearn.model_selection import train_test_split

# 훈련용 데이터와 평가용 데이터 분할 
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.30, random_state=0)

##2.3 원하는 모델을 이용하여 분석 모델 구축하기

* 다양한 모델을 사용하여 학습하기

In [744]:
from sklearn.naive_bayes import GaussianNB

model = GaussianNB().fit(X_train, Y_train)
Y_predict = model.predict(X_test)

In [677]:
from sklearn.neighbors import KNeighborsClassifier

kNN_model = KNeighborsClassifier(n_neighbors=5).fit(X_train, Y_train)
kNN_Y_predict = kNN_model.predict(X_test)

In [678]:
# from sklearn.svm import SVC

# knn_models = KNeighborsClassifier().fit(X_train, Y_train)
# knn_Y_predict = knn_models.predict(X_test)

# param= {'n_neighbors' : [2,3,4,5,6,7,8]}

# knn_grid = GridSearchCV(KNeighborsClassifier(), param, refit=True, cv = 5, verbose=3)
# knn_grid.fit(X_train,Y_train)

In [679]:
# knn_grid_pred = knn_grid.predict(X_test)

In [680]:
from sklearn.tree import DecisionTreeClassifier

tree_model = DecisionTreeClassifier(max_depth = 4, min_samples_split= 2, random_state= 1).fit(X_train, Y_train)
tree_Y_predict = tree_model.predict(X_test)

In [681]:
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV

svm_model=SVC(gamma='auto').fit(X_train, Y_train)
svm_y_predict = svm_model.predict(X_test)

param_grid = {'C': [0.01, 0.1, 1, 10, 100, 1000],
              'gamma': [10, 1, 0.1, 0.01, 0.001, 0.0001],
              'kernel': ['rbf']} 
 
grid = GridSearchCV(SVC(), param_grid, refit=True, cv = 5, verbose=3)
grid.fit(X_train,Y_train)

Fitting 5 folds for each of 36 candidates, totalling 180 fits
[CV 1/5] END ......C=0.01, gamma=10, kernel=rbf;, score=0.400 total time=   0.0s
[CV 2/5] END ......C=0.01, gamma=10, kernel=rbf;, score=0.400 total time=   0.0s
[CV 3/5] END ......C=0.01, gamma=10, kernel=rbf;, score=0.400 total time=   0.0s
[CV 4/5] END ......C=0.01, gamma=10, kernel=rbf;, score=0.400 total time=   0.0s
[CV 5/5] END ......C=0.01, gamma=10, kernel=rbf;, score=0.375 total time=   0.0s
[CV 1/5] END .......C=0.01, gamma=1, kernel=rbf;, score=0.400 total time=   0.0s
[CV 2/5] END .......C=0.01, gamma=1, kernel=rbf;, score=0.400 total time=   0.0s
[CV 3/5] END .......C=0.01, gamma=1, kernel=rbf;, score=0.400 total time=   0.0s
[CV 4/5] END .......C=0.01, gamma=1, kernel=rbf;, score=0.400 total time=   0.0s
[CV 5/5] END .......C=0.01, gamma=1, kernel=rbf;, score=0.375 total time=   0.0s
[CV 1/5] END .....C=0.01, gamma=0.1, kernel=rbf;, score=0.400 total time=   0.0s
[CV 2/5] END .....C=0.01, gamma=0.1, kernel=rbf

GridSearchCV(cv=5, estimator=SVC(),
             param_grid={'C': [0.01, 0.1, 1, 10, 100, 1000],
                         'gamma': [10, 1, 0.1, 0.01, 0.001, 0.0001],
                         'kernel': ['rbf']},
             verbose=3)

In [682]:
grid_pred_ = grid.predict(X_test)

##2.4 생성한 모델의 성능 확인하기

* accuracy 를 출력하세요. 


In [683]:
print(classification_report(Y_test, Y_predict))

              precision    recall  f1-score   support

           0       0.95      1.00      0.97        19
           1       1.00      0.91      0.95        22
           2       0.93      1.00      0.96        13

    accuracy                           0.96        54
   macro avg       0.96      0.97      0.96        54
weighted avg       0.97      0.96      0.96        54



In [684]:
print(classification_report(Y_test, kNN_Y_predict))
# print(classification_report(Y_test, knn_grid_pred))

              precision    recall  f1-score   support

           0       0.80      0.84      0.82        19
           1       0.77      0.77      0.77        22
           2       0.50      0.46      0.48        13

    accuracy                           0.72        54
   macro avg       0.69      0.69      0.69        54
weighted avg       0.72      0.72      0.72        54



In [685]:
print(classification_report(Y_test, tree_Y_predict))

              precision    recall  f1-score   support

           0       1.00      0.95      0.97        19
           1       0.95      0.95      0.95        22
           2       0.93      1.00      0.96        13

    accuracy                           0.96        54
   macro avg       0.96      0.97      0.96        54
weighted avg       0.96      0.96      0.96        54



In [686]:
print(classification_report(Y_test, grid_pred_))

              precision    recall  f1-score   support

           0       1.00      0.84      0.91        19
           1       0.85      1.00      0.92        22
           2       1.00      0.92      0.96        13

    accuracy                           0.93        54
   macro avg       0.95      0.92      0.93        54
weighted avg       0.94      0.93      0.93        54



In [687]:
# model accuarcy by Bayes
accuracy = accuracy_score(Y_test, Y_predict)
print("Accuracy: {0:.3f}".format(accuracy))

Accuracy: 0.963


In [688]:
# model accuarcy by Decision tree 
accuracy = accuracy_score(Y_test, tree_Y_predict)
print("Accuracy: {0:.3f}".format(accuracy))

Accuracy: 0.963


##2.5 채점 기준

- Accuracy 점수가 95 이상: **50점 (만점)**
- Accuracy 점수가 93 이상, 95미만: **45점**
- Accuracy 점수가 90 이상, 93미만: **40점**
- Accuracy 점수가 90 미만: **35점**